### DISCUSSION TITLE
Grafana can't connect to Supabase because of missing options field

### DISCUSSION THREAD

**Discussion by robertaboukhalil** [supabase/supabase#21060] [Questions]
When I create a Postgres data source on Grafana that connects to Supabase, I can't specify a connection string. There is a form for entering the host URL, database name, username, **but** there's no place for `options` = `reference=project_id`. Specifying the username as `postgres.project_id` doesn't seem enough.

Has anyone else run into this issue?

**Comment by mwoss** [comment]
Supabase exposes Prometheus compatible metrics for every project. Couldn't you directly ingest them directly in your Grafana? :D 
More info here - https://supabase.com/docs/guides/platform/metrics

**Reply by robertaboukhalil** [reply]
Thanks! In this case, I'm not using Grafana for infra metrics, but app-level metrics that would not be captured by prometheus.

**Reply by mwoss** [reply]
What do you mean by app-level metrics? Supabase exposes both Postgres and Dashboard-related metrics. By app-level I understand metrics generated by your own backend or something like that.

Btw here [supabase-grafana](https://github.com/supabase/supabase-grafana/blob/main/docs/metrics.md) you can find all metrics exposed by Supabase.

**Reply by robertaboukhalil** [reply]
Yes, by app level metrics, I mean I run SQL queries on my postgres tables to get custom metrics. This is for a tutorial website, so I calculate things like for a given tutorial, % of users that finish step 1, 2, etc. Whereas the metrics supabase-grafana provides are low-level metrics about CPU/disk/RAM/etc usage.

**Reply by mwoss** [reply]
Oh, got it, so you want to display analytical data there. Sorry, I misunderstood you.

What error are you getting when you set up a data source as described above? Did you take those values from  `Setting > Database > Connection parameters`? Have you tried without connection pooling via IPv4 (this method will be deprecated soon, but worth trying)?

**Reply by robertaboukhalil** [reply]
Correct! I get a "wrong password" error but I triple checked it's the right password and reentered it several times. I think the error is caused by not specifying `reference=project_id` in the options field since that field is not exposed in Grafana. Unfortunately ipv4 is already deprecated.

**Reply by mwoss** [reply]
Hmm 🤔 Have you tried to connect to Postgres using the same credentials set but via `psql`, just to verify they are correct?

**Reply by robertaboukhalil** [reply]
Good idea! It works with psql just fine, and I can reproduce the error if I don't specify options as `reference=project_id`. So since Grafana doesn't support the options field, I don't think this is supported unless I modify Grafana itself to give it a connection string directly.

**Reply by mwoss** [reply]
Make sense, sad it does not work then. I guess you could file an future request on Grafana repo, so they could implement it in the future :D

**Comment by AndersonMamede** [comment]
@robertaboukhalil  I've been using Grafana cloud to query the data from my Supabase project's database and I had the same issue you just reported. I tried to update the connection in Grafana to use the new connection pooling but it wasn't working and for a couple of hours I even broke the old connection that had been working for months and I couldn't make it work again. After spending a couple more hours trying different things, I finally have it working with the new connection pooling. This is the connection config that is working for me:

- Host URL: the host and port for the connection pooling (get it in Supabase's dashboard), e.g. `aws-0-us-west-1.pooler.supabase.com:5432`
- Database name: `postgres`
- Username: `postgres.<project_id>`
- Password: `<password>`

Another problem I experienced was that I actually use a custom user with read-only rights and I was getting the connection error `Tenant or user not found` even when trying locally with `psql`, but then I figured out that I can just set up the `Username` field like `<custom_username>.<project_id>` just like the default username `postgres.<project_id>` and then the connection worked :tada:

**Reply by hypercomputer** [reply]
Thanks so much for this, this comment saved me a ton of time.

**Reply by MichaelMilstead** [reply]
Thank you!
